python/source/flight.rst

```diff
@@ -752,4 +752,210 @@ Running the client script, you should see the server printing out information ab
+
+    class TestSpanExporter(SpanExporter):
+        def __init__(self):
+            self.spans = []
+    
+        def export(self, spans):
+            self.spans.extend(spans)
+            return SpanExportResult.SUCCESS
+
+    def configure_tracing():
+        # Service name is required for most backends,
+        # and although it's not necessary for console export,
+        # it's good to set service name anyways.
+        resource = Resource(attributes={
+            SERVICE_NAME: "my-service"
+        })
+        exporter = TestSpanExporter()
+        provider = TracerProvider(resource=resource)
+        processor = SimpleSpanProcessor(exporter)
+        provider.add_span_processor(processor)
+        trace.set_tracer_provider(provider)
+        return exporter
+
+We can use the middleware now in our EchoServer from earlier. 
+
+.. code-block::
+
+    if __name__ == '__main__':
+        exporter = configure_tracing()
+        server = EchoServer(
+            location="grpc://0.0.0.0:8816",
+            middleware={
+                "tracing": ServerTracingMiddlewareFactory()
+            },
+        )
+        server.serve()
+
+.. testcode::
+    :hide:
+
+    # Code block to start for real a server in background
+    # and wait for it to be available.
+    # Previous code block is just to show to user how to start it.
+    import threading
+    exporter = configure_tracing()
+    server = EchoServer(
+        location="grpc://0.0.0.0:8816",
+        middleware={
+            "tracing": ServerTracingMiddlewareFactory()
+        },
+    )
+    t = threading.Thread(target=server.serve)
+    t.start()
+
+Finally, we can use the middleware on the client. When we make a call with our 
+client within an OpenTelemetry span, our client middleware will create
+a child span for the client-side Flight call and then propagate the span
+context to the server. Our server middleware will pick up that trace context and create
+another child span.
+
+.. testcode::
+
+   import pyarrow as pa
+   import pyarrow.flight
+   from opentelemetry import trace
+
+   # Client would normally also need to configure tracing, but for this example
+   # the client and server are running in the same Python process.
+   # exporter = configure_tracing()
+
+   tracer = trace.get_tracer(__name__)
+
+   client = pa.flight.connect(
+       "grpc://0.0.0.0:8816",
+       middleware=[ClientTracingMiddlewareFactory()],
+   )
+
+   with tracer.start_as_current_span("hello_world") as span:
+       action = pa.flight.Action("echo", b"Hello, world!")
+       # Call list() on do_action to drain all results.
+       list(client.do_action(action=action))
+    
+   print(f"There are {len(exporter.spans)} spans.")
+   print(f"The span names are:\n  {list(span.name for span in exporter.spans)}.")
+   print(f"The span status codes are:\n  "
+         f"{list(span.status.status_code for span in exporter.spans)}.")
```

Comment:
It may not be clear here, but the status codes here are OpenTelemetry statuses, and not gRPC or Flight statuses

Reply:
We set them in the middleware above, so I think that's self explanatory enough for now. I suppose a more full-featured middleware would also set the gRPC status in the middleware with semantic conventions, but that seemed like it would overcomplicate this.